llama_index/ingestion/pipeline.py

```diff
@@ -195,8 +264,84 @@ def run(
         if self.reader is not None:
             input_nodes += self.reader.read()
 
+        return input_nodes
+
+    def _handle_duplicates(self, nodes: List[BaseNode]) -> List[BaseNode]:
+        """Handle docstore duplicates by checking all hashes."""
+        assert self.docstore is not None
+
+        existing_hashes = self.docstore.get_all_document_hashes()
+        current_hashes = []
+        nodes_to_run = []
+        for node in nodes:
+            if node.hash not in existing_hashes and node.hash not in current_hashes:
+                self.docstore.add_documents([node])
+                self.docstore.set_document_hash(node.id_, node.hash)
+                nodes_to_run.append(node)
+                current_hashes.append(node.hash)
+        return nodes_to_run
+
+    def _handle_upserts(self, nodes: List[BaseNode]) -> List[BaseNode]:
```

Comment:
for _handle_upserts is this function scope only for ingestion pipeline.
if a user not using ingestion pipeline, can they do upsert easily for vector/doc store

Reply:
Not sure what you mean by this 🤔